builder/build_validate.go

```diff
@@ -0,0 +1,927 @@
+		default:
+			tmp = append(tmp, f)
+		}
+	}
+	results.bundleDiff["os-core"].manFileDiffs.modList = tmp
+
+	if mandatoryFiles != 3 {
+		errorList = append(errorList, "ERROR: os-core is missing mandatory modified file(s)\n")
+	}
+
+	// Compare manifest files against package files. Inconsistencies are added to the errorList.
+	for _, b := range results.bundleDiff {
+		errorList = append(errorList, diffResultLists(b.pkgFileDiffs.addList, b.manFileDiffs.addList, toInfo[b.name].subPkgFiles, b.name, "added")...)
+		errorList = append(errorList, diffResultLists(b.pkgFileDiffs.modList, b.manFileDiffs.modList, toInfo[b.name].subPkgFiles, b.name, "modified")...)
+
+		// Added bundles will not exist in the fromInfo object and added bundles cannot have deleted files.
+		if fromInfo[b.name] != nil {
+			errorList = append(errorList, diffResultLists(b.pkgFileDiffs.delList, b.manFileDiffs.delList, fromInfo[b.name].subPkgFiles, b.name, "deleted")...)
+		}
+	}
+
+	return errorList, nil
+}
+
+// diffResultLists compares manifest files against package files and generates
+// an error string when they don't match.
+func diffResultLists(pkgFiles, manFiles []string, info map[string]*fileInfo, bundle, mode string) []string {
+	sort.Strings(pkgFiles)
+	sort.Strings(manFiles)
+
+	var i, j int
+	var errorList = []string{}
+
+	// Compare manifest and package file lists. When they don't match, add an
+	// error string to the error slice.
+	for i, j = 0, 0; i < len(pkgFiles) && j < len(manFiles); {
+		switch strings.Compare(pkgFiles[i], manFiles[j]) {
+		case 1:
+			// File in manifest list, but not package
+			errorMsg := fmt.Sprintf("ERROR: %s is %s in manifest '%s', but not in a package\n", manFiles[j], mode, bundle)
+			errorList = append(errorList, errorMsg)
+			j++
+		case -1:
+			// File in package list, but not manifest
+			errorMsg := fmt.Sprintf("ERROR: %s is %s in package '%s', but not %s in manifest '%s'\n", pkgFiles[i], mode, info[pkgFiles[i]].pkg, mode, bundle)
+			errorList = append(errorList, errorMsg)
+			i++
+		case 0:
+			i++
+			j++
+		}
+	}
+
+	for ; i < len(pkgFiles); i++ {
+		errorMsg := fmt.Sprintf("ERROR: %s is %s in package '%s', but not %s in manifest '%s'\n", pkgFiles[i], mode, info[pkgFiles[i]].pkg, mode, bundle)
+		errorList = append(errorList, errorMsg)
+	}
+	for ; j < len(manFiles); j++ {
+		errorMsg := fmt.Sprintf("ERROR: %s is %s in manifest '%s', but not in a package\n", manFiles[j], mode, bundle)
+		errorList = append(errorList, errorMsg)
+	}
+
+	return errorList
+}
+
+// printMcaResults displays any MCA errors and prints bundle diff statistics when there are no errors.
+func printMcaResults(results *mcaDiffResults, fromInfo, toInfo map[string]*mcaBundleInfo, fromVer, toVer int, errorList []string) error {
+	var err error
+
+	// An overwhelming number of errors can be generated when this test
+	// identifies a manifest bug, so limit the error output to 50.
+	if len(errorList) > 0 {
+		for i, err := range errorList {
+			if i == 50 {
```

Comment:
Would be nice to print a warning here to let the user know that there were errors that got omitted.